# Overview

There are 3 pipelines that come together for SNP calling:

1. Workflow 1 - BAM alignments
2. Workflow 2 - nQuire ploidy analysis
3. Workflow 3 - SNP calling

Each workflow has its own Snakefile and config.yaml. Workflow 1 and its Snakefile/config.yaml live in the project directory. Workflow 2 lives in the nQuire/ directory, and Workflow 3 in the snp-calling/ directory. These 2 directories must be created manually and the Snakefile and config.yaml placed into them manually.

NOTE: The pipeline currently does not remove asterisks from fasta before calculating snp distances! SEQKIT_REPLACE step from MycoSNP is not yet implemented!

# Workflow steps

## Workflow 1 - BAM alignments

1. Prepare reference
	1.1 Run nucmer on reference genome to identify repeats (nucmer)
	1.2 Generate repeat coordinates (show-coords)
	1.3 Convert repeat coordinates to bed file (awk)
	1.4 Mask reference genome using bedtools (maskfasta)
	1.5 Index reference with bwa-mem2 (index)
	1.6 Index reference with samtools (faidx)
	1.7 Create Picard sequence dictionary (CreateSequenceDictionary)

2. Align reads
	2.1 Align reads with bwa-mem2 (mem)
	2.2 Convert sam to bam and sort with samtools (view & sort)

3. Clean up alignments
	3.1 Mark duplicates with Picard (MarkDuplicates)
	3.2 Clean sam (bam) with Picard (CleanSam)
	3.3 Fix mate information with Picard (FixMateInformation)
	3.4 Replace read groups with Picard (AddOrReplaceReadGroups)

4. Generate alignment statistics
	4.1 Samtools coverage
	4.2 Samtools stats
	4.3 Samtools idx_stats
	4.4 Samtools flagstat

## Workflow 2 - nQuire ploidy analysis

1. Create .bin file (nQuire create)
2. Denoise .bin file (nQuire denoise)
3. Generate allele frequency histogram (nQuire histo)
4. Fit nQuire model (nQuire lrdmodel)
5. Summarize nQuire output (python script)

NOTE: If a bed file is provided (in the config), then the nQuire steps above will be repeated for each region (contig) in the bed file and a second summary file summarizing regions (contigs) produced as well.

## Workflow 3 - SNP calling

1. Call SNPs
	1.1 Generate GVCFs using GATK HaplotypeCaller
		- Allows specification of separate ploidy for each sample
	1.2 Combine gVCFs into single multi-sample GVCF (GATK CombineGVCFs)
	1.3 Perform joint genotyping (GATK GenotypeGVCFs)

2. Filter SNPs
	2.1 Filter SNPs based on quality (GATK VariantFiltration)
	2.2 Select only SNPs (GATK SelectVariants)
	2.3 Filter SNPs using on additional criteria (filterGatkGenotypes2.py)

3. Generate SNP distances
	3.1 Convert VCF to fasta (vcfToFasta-Broad-Edited2.py)
	3.2 Calculate SNP distances from fasta (snp-dists-Edited2.py)

# Required software

## Workflow 1

The following packages/tools must be in the path:
	Nucmer
	Bedtools
	bwa-mem2
	samtools

The path to the following packages/tools must be specified in the config file:
	picard.jar

## Workflow 2

The path to the following packages/tools must be specified in the config file:
	nQuire (directory containing executables)

## Workflow 3

The following packages/tools must be in the path:
	GATK

The path to the following packages/tools must be specified in the config file:
	filterGatkGenotypes2.py
	vcfToFasta-Broad-Edited2.py
	snp-dists-Edited2.py
	vcfTools2.py # Edited custom library required by above 3 scripts
	
NOTE: Each of the tools for each workflow above may have addtional dependencies; install as required and test tools work before running pipeline.

# Configuration files

An explanation of required arguments can be found in each workflow's config.yaml file.

# Execution order

Workflow 1 must be in and run from the project directory:

snakemake -s Snakefile --configfile config.yaml --cores 28 

Workflow 2 must be placed in the nQuire/ directory and run there:

snakemake -S Snakefile --configfile config.yaml --cores 4

Workflow 3 must be placed in the snp-calling/ directory and run from there:

snakemake -S Snakefile --configfile config.yaml --cores 28

# Required directory structure

## Workflow 1 before running:
project/
	Snakefile		# For Workflow 1
	config.yaml		# For Workflow 1

## Workflow 1 after running:
project/
	Snakefile
	config.yaml
	reference/	# Created by workflow
	samtools-stats/	# Created by workflow
	bam-alignments/	# Created by workflow
	
## Workflow 2 before running:

project/
	nQuire/
		Snakefile	# For Workflow 2
		config.yaml	# For Workflow 2
	bam-alignments/		# Generated by Workflow 1 - required!

## Workflow 2 after running:
project/
	nQuire/
		Snakefile
		config.yaml
		logs/				# Created by workflow
		nquire-bin/			# Created by workflow
		nquire-histo/			# Created by workflow
		nquire-lrd/			# Created by workflow
		nQuire_combined.tsv		# Created by workflow
		nQuire_combined_regions.tsv	# Created by workflow
		
## Workflow 3 before running:
project/
	snp-calling/
		Snakefile			# For Workflow 3
		config.yaml			# For Workflow 3
		filterGatkGenotypes2.py		# Required for Workflow 3
		snp-dists-Edited2.py		# Required for Workflow 3
		vcfToFasta-Broad-Edited2.py	# Required for Workflow 3
		vcfTools2.py			# Required for Workflow 3
	bam-alignments/				# Generated by Workflow 1 - required!

## Workflow 3 after running:
project/
	snp-calling/
		Snakefile
		config.yaml
		filterGatkGenotypes2.py
		snp-dists-Edited2.py
		vcfToFasta-Broad-Edited2.py
		vcfTools2.py
		combined/	# Created by workflow
		fasta/		# Created by workflow
		gvcfs/		# Created by workflow
		logs/		# Created by workflow
		vcfs/		# Created by workflow
